3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2017-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics
+package snowplow.enrich
+package common.enrichments.registry
+
+// Scala
+import org.json4s.JValue
+import scala.collection.JavaConverters._
+
+// Snowplow
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.jayway.jsonpath.MapFunction
+import com.snowplowanalytics.iglu.client.{SchemaCriterion, SchemaKey}
+import com.snowplowanalytics.snowplow.enrich.common.ValidatedNelMessage
+import com.snowplowanalytics.snowplow.enrich.common.utils.MapTransformer.TransformMap
+import com.snowplowanalytics.snowplow.enrich.common.utils.ScalazJson4sUtils
+
+// Scala libraries
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+
+// Java
+import java.security.{MessageDigest, NoSuchAlgorithmException}
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ObjectNode, TextNode}
+import com.github.fge.jsonschema.core.report.ProcessingMessage
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider
+import com.jayway.jsonpath.{Configuration, JsonPath => JJsonPath, Option => JOption}
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion
+
+// Scalaz
+import scalaz.Scalaz._
+import scalaz._
+
+/**
+ * PiiField trait. This corresponds to a configuration top-level field (i.e. either a POJO or a JSON field) along with
+ * a function to apply that strategy to a TransformMap
+ */
+sealed trait PiiField {
+  val strategy: PiiStrategy
+  val fieldName: String
+  def applyStrategy(fieldValue: String): String
+  val strategyApplication =
+    (v: Validation[String, String]) => v.flatMap(applyStrategy(_).success)
```

Comment:
Also, I don't see why this function is necessary. And even it is then it should be `protected`, because it is just implementation detail, nothing crucial for application's logic.